src/libos/src/poll/do_epoll/epoll_file.rs

```diff
@@ -227,6 +228,10 @@ impl EpollFile {
                     entry.reset_ready();
                 }
             }
+
+            if ready.len() == 0 || count_events >= max_events {
+                break;
+            }
```

Comment:
You are right that the original loop is missing the ending condition. But the condition is not `ready.len() == 0 || count_events >= max_events`. The reason is that `ready.len()` may never reach zero since some ready entries---those that are neither edge-triggered or one-shot---will be kept in the entry list (see line 223). So whether the ready list is empty or not does not matter when deciding when to stop.

I think this is the right way to terminate the loop.

```rust
let pop_count = (max_events - count_events).min(pop_quota);
if pop_count == 0 {
    return;
}
```

This is basically saying in natural language that "we want to return as many results as possible. so if we cannot pop any more candidates, then it is time to give up by exiting the loop."